media_driver/linux/common/os/i915/mos_bufmgr.c

```diff
@@ -4691,3 +4691,246 @@ mos_bufmgr_gem_init(int fd, int batch_size)
+    memclear(query);
+    query.num_items = 1;
+    query.items_ptr = (uintptr_t)&query_item;
+
+    ret = drmIoctl(fd, DRM_IOCTL_I915_QUERY, &query);
+    if (ret)
+    {
+        goto fini;
+    }
+
+    int i, num;
+    for (i = 0, num = 0; i < engines->num_engines; i++) {
+        struct drm_i915_engine_info *engine =
+            (struct drm_i915_engine_info *)&engines->engines[i];
+        if ( engine_class == engine->engine_class
+             && ((caps & engine->capabilities) || caps == 0))
+        {
+            ci->engine_class = engine_class;
+            ci->engine_instance = engine->engine_instance;
+            ci++;
+            num++;
+        }
+    }
+    if (num > *nengine)
+    {
+        fprintf(stderr,"%s: Number of engine instances out of range, %d,%d\n",
+                __FUNCTION__, num, *nengine);
+        goto fini;
+    }
+    *nengine = num;
+
+
+fini:
+    free(engines);
+    return ret;
+}
+
+int mos_set_context_param_load_balance(struct mos_linux_context *ctx,
+                     const struct i915_engine_class_instance *ci,
+                     unsigned int count)
+{
+    int ret;
+    uint32_t size;
+
+    assert(ci);
+
+    /*
+    struct i915_context_engines_load_balance {
+	struct i915_user_extension base;
+
+	__u16 engine_index;
+	__u16 num_siblings;
+	__u32 flags;
+
+	__u64 mbz64;
+
+	struct i915_engine_class_instance engines[count];
+    }balancer;
+    */
+    size = sizeof(struct i915_context_engines_load_balance) + count * sizeof(*ci);
+    uint32_t* balancer = (uint32_t*)malloc(size);
+    if (NULL == balancer)
+    {
+        return -ENOMEM;
+    }
+    memset(balancer, 0, size);
+    ((struct i915_context_engines_load_balance*) balancer)->base.name = I915_CONTEXT_ENGINES_EXT_LOAD_BALANCE;
+    ((struct i915_context_engines_load_balance*) balancer)->num_siblings = count;
+    memcpy(((struct i915_context_engines_load_balance*) balancer)->engines, ci, count * sizeof(*ci));
+
+    /*
+    struct i915_context_param_engines {
+    __u64 extensions;
+    struct i915_engine_class_instance engines[count];
+    }
+    */
+    size = sizeof(uint64_t) + (count + 1) * sizeof(*ci);
+    uint32_t* set_engines = (uint32_t*) malloc(size);
+    if (NULL == set_engines)
+    {
+        return -ENOMEM;
+    }
+
+    /* set_engines->extension */
+    ((struct i915_context_param_engines*) set_engines)->extensions = (uintptr_t)(balancer);
+
+    /* set_engines->engines[0],
+       KMD requires the first entry of the engine map to be:
+       class=I915_ENGINE_CLASS_INVALID:instance=I915_ENGINE_CLASS_INVALID_NONE
+       */
+    ((struct i915_context_param_engines*)set_engines)->engines[0].engine_class = I915_ENGINE_CLASS_INVALID;
+    ((struct i915_context_param_engines*)set_engines)->engines[0].engine_instance = I915_ENGINE_CLASS_INVALID_NONE;
+
+    /* set_engines->engines[1]..[count], the "real" engine map */
+    memcpy(&((struct i915_context_param_engines*)set_engines)->engines[1], ci, count * sizeof(*ci));
```

Comment:
Are you using map entries other than the VENG during submission? If not I'd recommend not configuring them and just have the VENG in the map.

Reply:
yeah, fixed.